fish-rust/src/common.rs

```diff
@@ -1143,8 +1144,19 @@ pub fn wcs2osstring(input: &wstr) -> OsString {
     OsString::from_vec(result)
 }
 
-pub fn wcs2zstring(input: &wstr) -> CString {
-    if input.is_empty() {
+/// Converts a Unicode string into a [`CString`], taking into account the user's current locale. As
+/// our wide string input may represent non-unicode characters or else Unicode characters falling
+/// into the private-use range, we shift these into fish's own private use range to make
+/// round-tripping through a wide string a lossless operation. `wcs2zstring()` undoes that.
+///
+/// While this is ostensibly for wide string inputs, we look at the input string char-by-char and to
+/// perform this operation we necessarily allocate, so this can actually be used for any type that
+/// can represent a sequence of characters.
+pub fn wcs2zstring<IntoChars>(input: IntoChars) -> CString
```

Comment:
This function should *not* be generic over character sequences.

The only thing this function does is reverse the effects of `str2wcstring`, and add a nul terminator. If the argument did not go through `str2wcstring` (or is a widestring literal), then it should not be passed to `wcs2zstring`. And `str2wcstring` gives us a `WString` so this should accept a `&wstr`; any other argument is likely a latent bug.

This should continue to take a `&wstr` (that also allows you to trivially implement the empty input optimization if you want).